[Merged by Bors] - chore(*): bump to lean 3.19.0c, fin is now a subtype

jcommelin opened the pull request:
* Some `decidable.*` order theorems have been moved to core.

* `fin` is now a subtype. 
This means that the whnf of `fin n` is now `{i // i < n}`.
Also, the coercion `fin n → nat` is now preferred over `subtype.val`.
The entire library has been refactored accordingly. (Although I probably missed some cases.)

* `in_current_file'` was removed since the bug in 
`in_current_file` was fixed in core.

* The syntax of `guard_hyp` in core was changed from
`guard_hyp h := t` to `guard_hyp h : t`, so the syntax
for the related `guard_hyp'`, `match_hyp` and
`guard_hyp_strict` tactics in `tactic.interactive` was changed as well.

---
<!-- put comments you want to keep out of the PR commit here -->

jcommelin committed:
Order stuff has now been declared in core

jcommelin committed:
data/fin is now working again

jcommelin committed:
fix set_theory/pgame

jcommelin committed:
fix data/equiv/fin

jcommelin committed:
fix order/rel_iso

jcommelin committed:
fix data/list/of_fn

jsm28 commented on src/data/fin.lean:
```diff
@@ -169,15 +182,15 @@ end bit
 whose value is the original number.  -/
 lemma coe_val_of_lt {n : ℕ} {a : ℕ} (h : a < n + 1) :
-  (a : fin (n + 1)).val = a :=
+  ((a : fin (n + 1)) : ℕ) = a :=
```
This change makes `coe_val_of_lt` the same as `coe_coe_of_lt`, meaning one of those two lemmas can be removed.

jcommelin replied:
I undid this change. It might be useful to keep the lemma.

jsm28 commented on src/data/fin.lean:
```diff
@@ -959,7 +972,7 @@ end fin
+⟨coe, subtype.coe_injective⟩
 
 -- Once lean#359 is fixed (making `fin n` a subtype), this can go away
 -- as a duplicate of `function.embedding.coe_subtype`.
```
These two definitions (which I never ended up using) can, as the comments on them note, be removed now that `fin` is a subtype.

jcommelin replied:
Removed these defs.

jcommelin committed:
fix data/vector2

jcommelin committed:
fix data/array/lemmas

rwbarton commented on src/set_theory/pgame.lean:
```diff
@@ -124,7 +125,7 @@ Construct a pre-game from list of pre-games describing the available moves for L
 -- TODO define this at the level of games, as well, and perhaps also for finsets of games.
 def of_lists (L R : list pgame.{0}) : pgame.{0} :=
-pgame.mk (fin L.length) (fin R.length) (λ i, L.nth_le i.val i.is_lt) (λ j, R.nth_le j.val j.is_lt)
+pgame.mk (fin L.length) (fin R.length) (λ i, L.nth_le i.val i.2) (λ j, R.nth_le j.val j.is_lt)
```
I think you want to undo this change, having added the import.

jcommelin replied:
Thanks, fixed.

rwbarton commented on src/data/vector2.lean:
```diff
@@ -183,15 +183,19 @@ lemma remove_nth_insert_nth_ne {v : vector α (n+1)} :
       have j_pos : 0 < j := lt_of_le_of_lt (zero_le i) h,
-      simp [h, h_nji, fin.lt_iff_val_lt_val, -subtype.val_eq_coe],
-      rw [show j.pred = j - 1, from rfl, list.insert_nth_remove_nth_of_ge, nat.sub_add_cancel j_pos],
+      rw [dif_neg], swap, { exact h_nji },
```
Isn't this just `rw dif_neg h_nji`, and so on?

jcommelin replied:
@rwbarton Nope, unfortunately not. There's a coe in the `dite`, and it doesn't want to simp away for some reason.

robertylewis replied:
`split_ifs` doesn't clean this up better, does it? This is pretty ugly :(

rwbarton commented on src/data/equiv/fin.lean:
```diff
@@ -56,12 +56,12 @@ def sum_fin_sum_equiv : fin m ⊕ fin n ≃ fin (m + n) :=
   left_inv := λ x, begin
     cases x with y y,
-    { simp [y.2, fin.ext_iff] },
+    { simp [y.2, fin.ext_iff, y.is_lt], },
```
Aren't `y.2` and `y.is_lt` the same thing?

jcommelin replied:
`y.is_lt` uses the coe.

rwbarton replied:
Do you still need `y.2` though? I think I will need to edit this particular function soon anyways, so I guess I don't really care.

jcommelin replied:
Fixed, good catch.

rwbarton commented on src/data/array/lemmas.lean:
```diff
@@ -153,13 +153,13 @@ begin
-  by_cases ij : i.1 = j,
-  { subst j, rw [show fin.mk i.val h₃ = i, from fin.eq_of_veq rfl,
+  by_cases ij : (i : ℕ) = j,
+  { subst j, rw [show (⟨(i : ℕ), h₃⟩ : fin _) = i, from fin.eq_of_veq rfl,
```
Doesn't `fin.mk` still exist? This is a bit ugly.

jcommelin replied:
`fin.mk` exists. Maybe it should be made reducible? Right now this line was broken.

jcommelin committed:
fix data/hash_map

jcommelin committed:
fix finset/basic

jcommelin committed:
fix finset/sort

rwbarton commented on src/data/finset/sort.lean:
```diff
@@ -202,11 +202,11 @@ only if `i = j`. Since they can be defined on a priori not defeq types `fin k` a
 necessarily `k = l`), the conclusion is rather written `i.val = j.val`. -/
```
No, it's not.

jcommelin replied:
Good catch. Fixed.

jcommelin committed:
fix data/bitvec/basic and lint data/fin

jcommelin committed:
add simp-lemma fin.mk_coe

jcommelin committed:
undo change in pgame

jcommelin committed:
fix simp in data/equiv/fin

jcommelin committed:
fix docstring in data/finset/sort

jcommelin committed:
hackfix set_theory/game/short

jcommelin committed:
fix data/fintype/card

jcommelin committed:
fix combinatorics/composition

jcommelin committed:
fix set_theory/cardinal

jcommelin committed:
fix computability/primrec

jcommelin committed:
fix data/finsupp/basic

jcommelin committed:
fix data/zmod/basic

jcommelin committed:
fix algebra/linear_recurrence

jcommelin committed:
fix topology/metric_space/gromov_hausdorff

bryangingechen commented on src/topology/metric_space/gromov_hausdorff.lean:
```diff
@@ -639,9 +639,9 @@ begin
-      let i := ((E q) ⟨y, ys⟩).1,
-      let hi := ((E q) ⟨y, ys⟩).2,
-      have ihi_eq : (⟨i, hi⟩ : fin (N q)) = (E q) ⟨y, ys⟩, by rw fin.ext_iff,
+      let i : ℕ := ((E q) ⟨y, ys⟩),
```
```suggestion
      let i : ℕ := (E q) ⟨y, ys⟩,
```
Lots of extra parentheses in this section.

jcommelin replied:
Good catch. Fixed.

bryangingechen committed:
remove in_current_file' since in_current_file was fixed in core

jcommelin committed:
fix analysis/analytic/composition

jcommelin committed:
fix some parens in topology/metric_space/gromov_hausdorff

jcommelin committed:
fix group_theory/perm/sign

jcommelin committed:
fix topology/list

jcommelin committed:
lint data/fin

bryangingechen committed:
change syntax of guard_hyp', match_hyp and guard_hyp_strict to match new syntax of guard_hyp

bryangingechen committed:
fix guard_hyp tests

bryangingechen committed:
more fixes

bryangingechen committed:
Merge remote-tracking branch 'community/master' into fix-fin

jcommelin committed:
Merge branch 'master' into fix-fin

robertylewis committed:
fix archive proof

This is a symptom of something worse: there's a clash between subtype dec_eq and fin decidable_eq

bryangingechen commented on archive/100-theorems-list/73_ascending_descending_sequences.lean:
```diff
@@ -104,7 +104,8 @@ begin
         -- It ends at `j` since `i < j`.
-        { rw [max_insert, ht₁.2.1, option.lift_or_get_some_some, max_eq_left, with_top.some_eq_coe],
+        { convert @max_insert _ _ j t using 4,
+          rw [ ht₁.2.1, option.lift_or_get_some_some, max_eq_left, with_top.some_eq_coe],
```
```suggestion
          rw [ht₁.2.1, option.lift_or_get_some_some, max_eq_left, with_top.some_eq_coe],
```

bryangingechen committed:
Update archive/100-theorems-list/73_ascending_descending_sequences.lean

robertylewis committed:
slightly nicer
